File: leetcode/238/ProductExceptSelf.java
import java.util.*;

class ProductExceptSelf {
    public static void main(String[] args) {
        Solution sl = new Solution();
        sl.productExceptSelf(new int[] { 2, 3, 4 });
    }
}

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int left = 1;
        int right = 1;
        int[] product = new int[nums.length];
        
        for (int i = 0; i < nums.length; i++) {
            product[i] = left;
            left *= nums[i]; 
        }
        
        for (int i = nums.length - 1; i > -1; i--) {
            product[i] = right * product[i]; // left * right
            right *= nums[i];
        }
        return product;
    }
}

File: leetcode/238/ProductExceptSelf.rb
def product_except_self(nums)
    left = 1;
    right = 1;
    product = [];
    
    puts product.length
    # for(let i = 0; i < nums.length; i++)
    for i in (0..nums.length - 1)
        product[i] = left;
        left = left*nums[i]; 
    end
    puts product
    # for(let i = nums.length -1; i > -1; i--)
    for i in ((nums.length - 1).downto(0))
    
        product[i] = right*product[i];
        right = right*nums[i];             
    end
    puts product
    product;
end

puts product_except_self([1,2,3,4]);